storage/src/mmr/mem.rs

```diff
@@ -374,15 +376,47 @@ impl<H: CHasher> Mmr<H> {
         self.dirty_nodes.clear();
         nodes.sort_by(|a, b| a.1.cmp(&b.1));
 
+        let mut same_height = Vec::new();
+        let mut current_height = 1;
         for (pos, height) in nodes {
-            let left = pos - (1 << height);
-            let right = pos - 1;
-            let digest = hasher.node_digest(
-                pos,
-                self.get_node_unchecked(left),
-                self.get_node_unchecked(right),
-            );
-            let index = self.pos_to_index(pos);
+            if height == current_height {
+                same_height.push(pos);
+                continue;
+            }
+            self.update_node_digests(&same_height, current_height);
+            same_height.clear();
+            current_height += 1;
+            same_height.push(pos);
+        }
+        self.update_node_digests(&same_height, current_height);
+    }
+
+    /// Update digests of the given set of nodes of equal height in the MMR. Since they are all at
+    /// the same height, this can be done in parallel.
+    fn update_node_digests(&mut self, same_height: &Vec<u64>, height: u32) {
+        let two_h = 1 << height;
+        let computed_digests: Vec<(usize, H::Digest)> = same_height
```

Comment:
I think you need to do `pool.install(` around this to use a preconfigured thread pool? Not sure what this defaults to:

            let pool = create_pool(context.with_label("pool"), 4).unwrap();


            // Create a vector of numbers
            let v: Vec<_> = (0..10000).collect();


            // Use the thread pool to sum the numbers
            pool.install(|| {
                assert_eq!(v.par_iter().sum::<i32>(), 10000 * 9999 / 2);
            });

Reply:
As I understand it rayon has its own default pool that will be used. Is the concern that this is not trackable by our runtime?

Reply:
changed it so that the user has to pass in the threadpool (giving the caller the chance to set the preferred concurrency).